File: WebComponents/demos/vui-video/MFC/Vui.Video.cpp
#include "Vui.Video.h"

VuiVideo* vuiVideoInstance;

void VideoOnPropertyChange(IJSObject* AObj, IJSProperty* AProp) {
	if (vuiVideoInstance == NULL) {
		return;
	}
	std::wstring propName = JSROGetPropertyName(AProp);
	if ((propName == L"state") && (vuiVideoInstance->OnStateChanged != NULL)) {
		vuiVideoInstance->OnStateChanged(vuiVideoInstance);
	}
	if ((propName == L"position") && (vuiVideoInstance->OnPositionChanged != NULL)) {
		vuiVideoInstance->OnPositionChanged(vuiVideoInstance);
	}
	if ((propName == L"length") && (vuiVideoInstance->OnLengthChanged != NULL)) {
		vuiVideoInstance->OnLengthChanged(vuiVideoInstance);
	}
}


VuiVideo::VuiVideo()
{
	vuiVideoInstance = this;

	WCHAR appPath[MAX_PATH];
	GetModuleFileNameW(NULL, appPath, MAX_PATH);
	m_Xtagdir = appPath;
	const size_t p = m_Xtagdir.rfind('\\');
	if (p != std::string::npos) {
		m_Xtagdir = m_Xtagdir.substr(0, p);
	}
	m_Xtagdir += L"\\x-tag\\";

	m_Vui = new VirtualUI();
}

VuiVideo::~VuiVideo()
{
	vuiVideoInstance = NULL;
}

std::wstring VuiVideo::XTagDir() {
	return m_Xtagdir;
}

void VuiVideo::XTagDir(std::wstring value) {
	m_Xtagdir = value;
}

float VuiVideo::Position() {
	IJSProperty* prop = m_Video->Properties()->Item(L"position");
	return JSROGetPropAsFloat(prop);
}

void VuiVideo::Position(float value) {
	IJSProperty* prop = m_Video->Properties()->Item(L"position");
	JSROSetPropAsFloat(prop, value);
}

std::wstring VuiVideo::Src() {
	return m_Src;
}

void VuiVideo::Src(std::wstring value) {
	std::wstring src = value;
	
	//TODO: Check if is file in a better way
	bool isFile = (src.find('\\') != std::string::npos);
	if (isFile) {
		src = m_Vui->HTMLDoc()->GetSafeURL(src, 60);
	}

	IJSProperty* prop = m_Video->Properties()->Item(L"src");
	JSROSetPropAsString(prop, src);
}

float VuiVideo::Length() {
	IJSProperty* prop = m_Video->Properties()->Item(L"length");
	return JSROGetPropAsFloat(prop);
}

std::wstring VuiVideo::State() {
	IJSProperty* prop = m_Video->Properties()->Item(L"state");
	return JSROGetPropAsString(prop);
}

void VuiVideo::Play() {
	m_Video->Events()->Item(L"Play")->Fire();
}

void VuiVideo::Pause() {
	m_Video->Events()->Item(L"Pause")->Fire();
}

void VuiVideo::Stop() {
	m_Video->Events()->Item(L"Stop")->Fire();
}

void VuiVideo::Move(float position) {
	IJSEvent* evt = m_Video->Events()->Item(L"move");
	JSROSetEventArgAsFloat(evt, L"Position", position);
	evt->Fire();
}

void VuiVideo::CreateComponent(std::wstring name, INT64 containerHandle) {
	m_Vui->HTMLDoc()->CreateSessionURL(std::wstring(L"/x-tag/"), m_Xtagdir);
	m_Vui->HTMLDoc()->LoadScript(std::wstring(L"/x-tag/x-tag-core.min.js"), L"");
	m_Vui->HTMLDoc()->ImportHTML(std::wstring(L"/x-tag/vui-video/vui-video.html"), L"");
	m_Vui->HTMLDoc()->CreateComponent(name, std::wstring(L"vui-video"), containerHandle);

	m_Video = new JSObject(name);
	m_Video->OnPropertyChange = VideoOnPropertyChange;

	IJSProperty* prop;
	prop = m_Video->Properties()->Add(L"state");
	JSROSetPropAsString(prop, L"");

	prop = m_Video->Properties()->Add(L"position");
	JSROSetPropAsFloat(prop, 0);

	prop = m_Video->Properties()->Add(L"length");
	JSROSetPropAsFloat(prop, 0);

	prop = m_Video->Properties()->Add(L"src");
	JSROSetPropAsString(prop, L"");

	IJSEvent* event = m_Video->Events()->Add(L"move");
	JSROAddEventArgument(event, L"position", JSDT_FLOAT);

	m_Video->Events()->Add(L"play");
	m_Video->Events()->Add(L"pause");
	m_Video->Events()->Add(L"stop");
	m_Video->ApplyModel();
}

File: WebComponents/x-tag/vui-video/Vui.Video.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.IO;
using Cybele.Thinfinity;

namespace Cybele.Thinfinity.WebComponents
{
    class VuiVideo
    {
        private string m_Xtagdir;
        private JSObject m_video;
        private String m_src;

        private VirtualUI vui;

        public VuiVideo()
        {
            vui = new VirtualUI();
            m_Xtagdir = AppDomain.CurrentDomain.BaseDirectory;
            DirectoryInfo di = new DirectoryInfo(m_Xtagdir);
            while (di!=null)
            {
                m_Xtagdir = di.FullName +@"\x-tag\";
                if (Directory.Exists(m_Xtagdir)) return;
                di = di.Parent;
            }
        }
        public string Dir
        {
            get
            {
                return new DirectoryInfo(AppDomain.CurrentDomain.BaseDirectory).ToString();
            }
        }
        public string XTagDir
        {
            get
            {
                return m_Xtagdir;
            }
            set {
                m_Xtagdir = value;
            }
        }
        public float Position {
            get
            {
                return m_video.Properties["position"].AsFloat;
            }
            set
            {
                m_video.Properties["position"].AsFloat = value;
            }
        }
        public float Length {
            get
            {
                return m_video.Properties["length"].AsFloat;
            }
        }
        public String State {
            get
            {
                return m_video.Properties["state"].AsString;
            }
        }
        public void Play()
        {
            m_video.Events["play"].Fire();
        }
        public void Pause()
        {
            m_video.Events["pause"].Fire();
        }
        public void Stop()
        {
            m_video.Events["stop"].Fire();
        }
        public void Move(float position)
        {
            m_video.Events["move"].ArgumentAsFloat("Position",position).Fire();
        }
        public String Src
        {
            get
            {
                return m_src;
            }
            set
            {
                m_src = value;
                var url = m_src;
                if (File.Exists(m_src))
                {
                    url = vui.HTMLDoc.GetSafeUrl(m_src, 60);
                }
                m_video.Properties["src"].AsString = url;
            }
        }

        public event EventHandler<EventArgs> OnStateChanged;
        public event EventHandler<EventArgs> OnLengthChanged;
        public event EventHandler<EventArgs> OnPositionChanged;

        public void CreateComponent(Control ctrl)
        {
            vui.HTMLDoc.CreateSessionURL("/x-tag/", m_Xtagdir);
            vui.HTMLDoc.LoadScript(@"/x-tag/x-tag-core.min.js");
            vui.HTMLDoc.ImportHTML(@"/x-tag/vui-video/vui-video.html");
            vui.HTMLDoc.CreateComponent(ctrl.Name, "vui-video", ctrl.Handle);

            // -- The given name, is how the model shown this object in the model reference.
            m_video = new JSObject(ctrl.Name);
            m_video.OnPropertyChange += m_video_OnPropertyChange;
            // -- Adding properties, methods and events.
            m_video.Properties.Add("state").AsString = "";
            m_video.Properties.Add("position").AsFloat = 0.0f;
            m_video.Properties.Add("length").AsFloat = 0.0f;
            m_video.Properties.Add("src").AsString = "";
            m_video.Events.Add("play");
            m_video.Events.Add("pause");
            m_video.Events.Add("stop");
            m_video.Events.Add("move").AddArgument("position",IJSDataType.JSDT_FLOAT);
            m_video.ApplyModel();


        }

        void m_video_OnPropertyChange(object sender, JSPropertyChangeEventArgs e)
        {
            if ((e.Prop.Name == "state") && (OnStateChanged != null)) OnStateChanged(this, new EventArgs());
            if ((e.Prop.Name == "position") && (OnPositionChanged != null)) OnPositionChanged(this, new EventArgs());
            if ((e.Prop.Name == "length") && (OnLengthChanged != null)) OnLengthChanged(this, new EventArgs());
        }
    }
}
